AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusVisualizer.java

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package com.microsoft.accessibilityinsightsforandroidservice;
+
+import android.view.accessibility.AccessibilityEvent;
+import android.view.accessibility.AccessibilityNodeInfo;
+import java.util.ArrayList;
+
+public class FocusVisualizer {
+  private ArrayList<FocusElementHighlight> focusElementHighlights;
+  private ArrayList<FocusElementLine> focusElementLines;
+  private int tabStopCount;
+  private FocusVisualizerStyles styles;
+  private FocusVisualizationCanvas focusVisualizationCanvas;
+
+  public FocusVisualizer(
+      FocusVisualizerStyles focusVisualizerStyles,
+      FocusVisualizationCanvas focusVisualizationCanvas) {
+    this.focusElementHighlights = new ArrayList<>();
+    this.focusElementLines = new ArrayList<>();
+    this.tabStopCount = 0;
+    this.styles = focusVisualizerStyles;
+    this.focusVisualizationCanvas = focusVisualizationCanvas;
+  }
+
+  public void HandleAccessibilityRedrawEvent(AccessibilityEvent event) {
+    if (event.getEventType() == AccessibilityEvent.TYPE_WINDOW_CONTENT_CHANGED) {
+      return;
+    }
+
+    if (event.getEventType() == AccessibilityEvent.TYPE_WINDOW_STATE_CHANGED) {
+      this.resetVisualizations();
+      return;
+    }
+
+    if (event.getEventType() == AccessibilityEvent.TYPE_VIEW_SCROLLED) {
+      this.updateDrawingsWithNewCoordinates();
+      return;
+    }
+    if (event.getEventType() == AccessibilityEvent.TYPE_WINDOWS_CHANGED) {
+      this.resetVisualizations();
+      return;
+    }
+  }
+
+  public void HandleAccessibilityFocusEvent(AccessibilityEvent event) {
+    tabStopCount++;
+
+    AccessibilityNodeInfo eventSource = event.getSource();
+    AccessibilityNodeInfo previousEventSource = this.getPreviousEventSource();
+
+    this.createFocusElementHighlight(eventSource);
+    this.createFocusElementLine(eventSource, previousEventSource);
+
+    if (this.focusElementHighlights.size() > 1) {
+      this.setPreviousElementHighlightNonCurrent(
+          this.focusElementHighlights.get(this.focusElementHighlights.size() - 2));
```

Comment:
It's not immediately apparent what's special about the value of 2, making it a magic number. Would it help code readability to use a named constant here?. Presumably the same constant would also be used in line 61.

Reply:
@DaveTryon this is a good suggestion, there's no reason that we can't create the new elements after we set the previous elements as non-current, which would allow us to use -1 instead of -2 to just get the last element in the current list.  I've updated the code to handle it this way!